xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+      checkState(requestWriter != null, "ADS stream has not been started");
+      String version = "";
+      String nonce = "";
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        version = ldsVersion;
+        nonce = ldsRespNonce;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        checkArgument(resourceNames.size() == 1,
+            "RDS request requesting for more than one resource");
+        version = rdsVersion;
+        nonce = rdsRespNonce;
+        rdsResourceName = resourceNames.iterator().next();
+      }
+      // TODO(chengyuanzhang): cases for CDS/EDS.
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(version)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Sends a DiscoveryRequest with the given information as an ACK. Updates the latest accepted
+     * version for the corresponding resource type.
+     */
+    private void sendAckRequest(String typeUrl, Collection<String> resourceNames,
+        String versionInfo, String nonce) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        ldsVersion = versionInfo;
+        ldsRespNonce = nonce;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        rdsVersion = versionInfo;
+        rdsRespNonce = nonce;
+      }
+      // TODO(chengyuanzhang): cases for CDS/EDS.
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Sends a DiscoveryRequest with the given information as an NACK. NACK takes the previous
+     * accepted version.
+     */
+    private void sendNackRequest(String typeUrl, Collection<String> resourceNames, String nonce,
+        String message) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      String versionInfo = "";
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        versionInfo = ldsVersion;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        versionInfo = rdsVersion;
+      }
+      // TODO(chengyuanzhang): cases for CDS/EDS.
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .setErrorDetail(com.google.rpc.Status.newBuilder().setMessage(message))
```

Comment:
You should also set the code in Status:
```
setErrorDetail(com.google.rpc.Status.newBuilder().setCode(code).setMessage(message))
```
and `code` should be a param of `sendNackRequest`. All callers should provide the code from google.rpc.code.proto (or `com.google.rpc.Code`).

Reply:
Added `Code`.